indexprotocol/votings/stakingprotocol.go

```
@@ -275,6 +292,64 @@ func (p *Protocol) getStakingBucketInfoByEpoch(height, epochNum uint64, delegate
 	return votinginfoList, nil
 }
 
+func (p *Protocol) getAllStakingDelegateRewardPortions(epochStartHeight, epochNumber uint64, chainClient iotexapi.APIServiceClient) (blockRewardPercentage, epochRewardPercentage, foundationBonusPercentage map[string]float64, err error) {
+	blockRewardPercentage = make(map[string]float64)
+	epochRewardPercentage = make(map[string]float64)
+	foundationBonusPercentage = make(map[string]float64)
+	if epochStartHeight == p.epochCtx.FairbankHeight() {
+		// init from contract,from contract deployed height to epochStartheight-1,get latest portion
+		if p.rewardPortionContract == "" {
+			// todo make sure if ignore this error
+			//err = errors.New("portion contract address is empty")
+			return
+		}
+		if p.rewardPortionContractDeployHeight > epochStartHeight {
+			// todo make sure if ignore this error
+			//err = errors.New("portion contract deploy height should less than fairbank height")
+			return
+		}
+		count := epochStartHeight - p.rewardPortionContractDeployHeight
+		blockRewardPercentage, epochRewardPercentage, foundationBonusPercentage, err = getlog(p.rewardPortionContract, p.rewardPortionContractDeployHeight, count, chainClient, p.abi)
+		if err != nil {
+			err = errors.Wrap(err, "get log from chain error")
+			return
+		}
+	} else {
+		// get from mysql first
+		blockRewardPercentage, epochRewardPercentage, foundationBonusPercentage, err = getLastEpochPortion(p.Store.GetDB(), epochNumber-1)
+		if err != nil && errors.Cause(err) != indexprotocol.ErrNotExist {
+			// todo make sure if ignore this error
```

ditto

---

I don't think we need to check errors.Cause(err) != indexprotocol.ErrNotExist here. If errors.Cause(err) = indexprotocol.ErrNotExist, it is unexpected and we should throw the error, too.